src/Chapter3.hs

```diff
@@ -479,6 +518,37 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data Castle = Castle { castleName :: String }
+
+data Wall = Wall | NoWall
+
+data EducationCentre = Church | Library
+
+data House = OnePerson | TwoPerson | ThreePerson | FourPerson
+
+data City
+  = CityWithCastle Castle Wall EducationCentre [House]
+  | CityWithoutCastle EducationCentre [House]
```

Comment:
Although this totally makes sense, it would be much easier to have `City` as a product type.
Let's see how it fits:

first, we need to decide with the castle. Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

 1. Castle with a name
 2. Castle with the name + walls
 3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Then let's build the city:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cityCastle :: Castle -- from the above
    -- Church or library but not both
    , cityChurchOrLibrary :: EducationCentre -- you have already created this one
    -- any number of houses. Each house has one, two, three or four __people__ inside.
    , cityHouses :: [House]
    }
```

And after that, you would be able to implement other functions easily. Let me know if this helps!